projects/openapi-utilities/src/diff/diff.ts

```diff
@@ -0,0 +1,209 @@
+import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
+import { getParameterIdentity, isParameterObject } from './array-identifiers';
+
+export type JSONValue =
+  | null
+  | string
+  | number
+  | boolean
+  | JSONObject
+  | JSONArray;
+
+export type JSONArray = JSONValue[];
+
+export type JSONObject = {
+  [x: string]: JSONValue;
+};
+
+export type ObjectDiff =
+  | {
+      // Added
+      before?: undefined;
+      after: string;
+      pathReconciliation?: undefined;
+    }
+  | {
+      // Changed
+      before: string;
+      after: string;
+      pathReconciliation?: undefined;
+    }
+  | {
+      // Removed
+      before: string;
+      after?: undefined;
+      // Required to maintain pointers from reordering array keys
+      pathReconciliation: [number, string][];
+    };
+
+type StackItem<T extends JSONObject | JSONArray = JSONObject | JSONArray> = [
+  {
+    value: T;
+    path: string;
+  },
+  {
+    value: T;
+    path: string;
+  }
+];
+
+// Diffs two objects, generating the leaf nodes that have changes
+export function diff(before: any, after: any): ObjectDiff[] {
+  const diffResults: ObjectDiff[] = [];
+  const stack: StackItem[] = [
+    [
+      { value: before, path: '' },
+      { value: after, path: '' },
+    ],
+  ];
+  while (stack.length > 0) {
+    const [before, after] = stack.pop()!;
+    const comparisons: {
+      beforeValue: JSONValue | undefined;
+      beforePath: string;
+      afterValue: JSONValue | undefined;
+      afterPath: string;
+    }[] = [];
+
+    // TODO in the future, skip adding comparisons based on diff preprocessing step
```

Comment:
When we are ready for this, we can modify this fn to take in merkel trees to optimize the speed of diff here